pallets/amm-support/src/lib.rs

```diff
@@ -0,0 +1,186 @@
+mod tests;
+
+// Re-export pallet items so that they can be accessed from the crate namespace.
+pub use pallet::*;
+
+pub const MAX_STACK_SIZE: u32 = 10;
+
+#[derive(RuntimeDebug, Encode, Decode, Default, Clone, PartialEq, Eq, TypeInfo, MaxEncodedLen)]
+pub struct ExecutionIdStack(BoundedVec<ExecutionType<IncrementalIdType>, ConstU32<MAX_STACK_SIZE>>);
+impl ExecutionIdStack {
+	fn push(&mut self, execution_type: ExecutionType<IncrementalIdType>) -> Result<(), ()> {
+		self.0.try_push(execution_type).map_err(|_| ())
+	}
+
+	fn pop(&mut self) -> Result<ExecutionType<IncrementalIdType>, ()> {
+		self.0.pop().ok_or(())
+	}
+
+	fn get(self) -> Vec<ExecutionType<IncrementalIdType>> {
+		self.0.into_inner()
+	}
+}
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+	use frame_support::pallet_prelude::*;
+
+	/// The current storage version.
+	const STORAGE_VERSION: StorageVersion = StorageVersion::new(1);
+
+	#[pallet::pallet]
+	#[pallet::storage_version(STORAGE_VERSION)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+	}
+
+	#[pallet::storage]
+	/// Next available incremental ID
+	#[pallet::getter(fn incremental_id)]
+	pub(super) type IncrementalId<T: Config> = StorageValue<_, IncrementalIdType, ValueQuery>;
+
+	#[pallet::storage]
+	/// Next available incremental ID
+	#[pallet::getter(fn id_stack)]
+	pub(super) type IdStack<T: Config> = StorageValue<_, ExecutionIdStack, ValueQuery>;
+
+	#[pallet::error]
+	pub enum Error<T> {
+		MaxStackSizeReached,
+		EmptyStack,
+	}
+
+	// on initialize - clear operation_id stack if not empty
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(crate) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// Trade executed.
+		Swapped {
+			swapper: T::AccountId,
+			filler: T::AccountId,
+			filler_type: Filler<AssetId, OtcOrderId>,
+			operation: TradeOperation,
+			inputs: Vec<(AssetType<AssetId, NftId>, Balance)>,
+			outputs: Vec<(AssetType<AssetId, NftId>, Balance)>,
+			fees: Vec<Fee<AssetId, Balance, T::AccountId>>,
+			operation_id: Vec<ExecutionType<IncrementalIdType>>,
+		},
+	}
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		fn on_initialize(_n: BlockNumberFor<T>) -> Weight {
+			let mut weight: Weight = Weight::zero();
+			weight.saturating_accrue(T::DbWeight::get().reads_writes(1, 1));
+
+			<Self as ExecutionTypeStack<IncrementalIdType>>::clear();
+
+			Weight::from_parts(weight.ref_time(), 0)
+		}
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {}
+}
+
+impl<T: Config> Pallet<T> {
+	/// Returns next incremental ID and updates the storage.
+	pub fn next_incremental_id() -> Result<IncrementalIdType, DispatchError> {
+		IncrementalId::<T>::try_mutate(|current_id| -> Result<IncrementalIdType, DispatchError> {
+			let inc_id = *current_id;
+			*current_id = current_id.checked_add(1).ok_or(ArithmeticError::Overflow)?;
```

Comment:
use overflowing. so when reaches the end - it starts from 0 rather than make all calls failing due to this.